odoo/addons/base/models/ir_model.py

```diff
@@ -1514,26 +1553,23 @@ def check(self, model, mode='read', raise_exception=True):
         self.flush(self._fields)
 
         # We check if a specific rule exists
-        self._cr.execute("""SELECT MAX(CASE WHEN perm_{mode} THEN 1 ELSE 0 END)
+        self._cr.execute("""SELECT coalesce(bool_or(perm_{mode}), false)
                               FROM ir_model_access a
                               JOIN ir_model m ON (m.id = a.model_id)
-                              JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
+                              LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
                              WHERE m.model = %s
-                               AND gu.uid = %s
+                               AND (a.group_id IS NULL OR gu.uid = %s)
                                AND a.active IS TRUE""".format(mode=mode),
                          (model, self._uid,))
         r = self._cr.fetchone()[0]
 
-        if not r:
-            # there is no specific rule. We check the generic rule
-            self._cr.execute("""SELECT MAX(CASE WHEN perm_{mode} THEN 1 ELSE 0 END)
-                                  FROM ir_model_access a
-                                  JOIN ir_model m ON (m.id = a.model_id)
-                                 WHERE a.group_id IS NULL
-                                   AND m.model = %s
-                                   AND a.active IS TRUE""".format(mode=mode),
-                             (model,))
-            r = self._cr.fetchone()[0]
+        # FIXME: if model not in self.env, it'll be missing from _acl as well
+        other = self._acl()[model]
+        if r != other[mode]:
+            _logger.warning(
+                "%s %s: expected %s got %s",
+                model, mode, r, other
+            )
```

Comment:
@odony question regarding modifications to `IrModelAccess.check`:  currently I've an assertion check there and I merged the two queries into one, but maybe there should just be _acl in stable, and then in the master forward-port I'll make IrModelAccess delegate to _acl instead of implementing the check internally?

Reply:
It might be possible to recode the entirety of _visible_menu_ids as a single query but it strikes me as pretty difficult. It's down to 56 queries but that could still be improved (42 out of the 56 come from calling exists() on individual actions)

Reply:
After some more messing about, `_visible_menu_ids` is down to 14 queries and 69 ±12ms, from 57 queries and 77 ±11ms. The gain is not great and the code looks pretty weird at this point.

Reply:
Definitely convoluted. The part I don't like is all this business to trick the ORM into batching queries, resolving the models for actions based our horrible legacy schema, checking they still exist, etc.

Most of the lines in `_visible_menu_ids` are dedicated to that, and it feels like this could be replaced with a direct query like this:

```sql
   SELECT m.id
     FROM ir_ui_menu m
     JOIN ir_actions a on (a.id = substring(m.action, strpos(m.action,',') + 1)::int)
LEFT JOIN ir_act_server s ON (a.id = s.id)
LEFT JOIN ir_act_window w ON (a.id = w.id)
LEFT JOIN ir_act_report_xml r ON (a.id = r.id)
LEFT JOIN ir_model model ON (model.id = s.model_id or model.model = coalesce(w.res_model, r.model))
     JOIN _acl ON (model.model = _acl.model)
    WHERE _acl.perm_read
```

This assumes that `_acl` is provided by a CTE equivalent to your `_acl()`. You could even skip the JOIN on `ir_model` then.
Even better: to keep the SQL simpler just use `WHERE model.model in %s`, and use `_acl()` to obtain the list of accessible models separately.

Wouldn't this get rid of all the weird ORM things in `_visible_menu_ids`, leaving on some minor details like the debug mode and parent menus to be dealt with, using straightforward ORM `search()` to obtain the relevant menu ids?

Reply:
FWIW the combined query I was using to test the approach is this:
```sql
WITH _acl (model, perm_read) AS (
       SELECT model, true FROM ir_model WHERE transient
       UNION ALL 
       SELECT m.model, coalesce(acl.perm_read, false)
         FROM ir_model m
    LEFT JOIN (
               SELECT a.model_id, bool_or(a.perm_read) AS perm_read
                 FROM ir_model_access a
            LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
                WHERE a.active AND (a.group_id IS NULL OR gu.uid = %s)
             GROUP BY a.model_id
    ) AS acl ON (acl.model_id = m.id)
        WHERE NOT m.transient
)
   SELECT m.id
     FROM ir_ui_menu m
     JOIN ir_actions a on (a.id = substring(m.action, strpos(m.action,',') + 1)::int)
LEFT JOIN ir_act_server s ON (a.id = s.id)
LEFT JOIN ir_act_window w ON (a.id = w.id)
LEFT JOIN ir_act_report_xml r ON (a.id = r.id)
LEFT JOIN ir_model model ON (model.id = s.model_id or model.model = coalesce(w.res_model, r.model))
LEFT JOIN _acl ON (model.model = _acl.model)
    WHERE _acl.perm_read or model.id IS NULL;
```

Reply:
> Wouldn't this get rid of all the weird ORM things in `_visible_menu_ids`, leaving on some minor details like the debug mode and parent menus to be dealt with, using straightforward ORM `search()` to obtain the relevant menu ids?

Might yes, especially retrieving the parent_id as well as the menu's own in order to build the final "visibility tree". The ACL bit could even remain outside of it, so this would just retrieve the (id, parent_id, model), replacing all the prefetch mess with a single query. Could work.

Reply:
@odony `_visible_menu_ids` is now down to 7 queries, though the performance changes with respect to `/web` are limited as it was already way down the profile.